lib/core/include/transport/default_transport.hpp

```diff
@@ -267,16 +373,25 @@ namespace irods::experimental::io::NAMESPACE_IMPL
 
             _func(input);
 
-            // TODO Modularize the block of code below.
-
-            const auto fd = rxDataObjOpen(comm_, &input);
+            // FIXME The memory referenced by this pointer will need to be freed
+            // by the shared library.
+            char* json_output{}; 
+            const auto fd = rx_replica_open(comm_, &input, &json_output);
```

Comment:
Need to think about this one.

Reply:
iRODS needs to provide its own heap memory deallocation function (i.e. `irods_free`). The client can't assume that memory allocated by the API is safe to be freed using `std::free`.

Reply:
Or the iRODS API must clearly state how allocated memory should be deallocated. This could be as simple as stating that all API functions are required to use `malloc` and `free` for memory allocation. This information should be documented in the header file of each API function.

Reply:
We had a head start with the life_time_manager library, is this still a place where we want to handle this?

Reply:
C clients can't use that library and that library doesn't exactly fix the issue of mixed memory allocation functions.  This is mostly about guaranteeing that users of the C API free memory correctly. The lifetime manager would use this C function underneath guaranteeing safe C++ code.

Reply:
this needs further investigation. leaving resolution of this issue out of this PR.